File: rest/admin-ui-ext/src/main/java/org/keycloak/admin/ui/rest/model/RoleMapper.java
package org.keycloak.admin.ui.rest.model;

import org.keycloak.models.ClientModel;
import org.keycloak.models.RealmModel;
import org.keycloak.models.RoleModel;

public class RoleMapper {
    public static ClientRole convertToModel(RoleModel roleModel, RealmModel realm) {
        ClientModel clientModel = realm.getClientById(roleModel.getContainerId());
        if (clientModel==null) {
            throw new IllegalArgumentException("Could not find referenced client");
        }
        ClientRole clientRole = new ClientRole(roleModel.getId(), roleModel.getName(), roleModel.getDescription());
        clientRole.setClientId(clientModel.getId());
        clientRole.setClient(clientModel.getClientId());
        return clientRole;
    }
}

File: keycloak-theme/src/main/java/org/keycloak/admin/ui/rest/model/RoleMapper.kt
package org.keycloak.admin.ui.rest.model

import org.keycloak.models.ClientModel
import org.keycloak.models.RoleModel
import org.mapstruct.*
import java.util.stream.Stream

@Mapper
abstract class RoleMapper {

    @Mapping(source = "name", target = "role")
    @Mapping(target = "client", ignore = true)
    abstract fun convertToRepresentation(role: RoleModel, @Context clientModel: Stream<ClientModel>): ClientRole

    @AfterMapping
    fun convert(role: RoleModel, @MappingTarget clientRole: ClientRole, @Context list: Stream<ClientModel>) {
        val clientModel = list.filter { c -> role.containerId == c.id }.findFirst().get()
        clientRole.clientId = clientModel.id
        clientRole.client = clientModel.clientId
    }
}